rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,556 @@
+}
+
+impl<G> TokenSwapper<G>
+where
+    G: NodeCount
+        + EdgeCount
+        + IntoEdges
+        + Visitable
+        + NodeIndexable
+        + IntoNeighborsDirected
+        + IntoNodeIdentifiers
+        + Debug,
+    G::NodeId: Hash + Eq + Debug,
+{
+    fn new(
+        graph: G,
+        mapping: HashMap<G::NodeId, G::NodeId>,
+        trials: Option<usize>,
+        seed: Option<u64>,
+    ) -> Self {
+        TokenSwapper {
+            graph,
+            input_mapping: mapping,
+            trials: trials.unwrap_or(4),
+            rng_seed: seed,
+            node_map: HashMap::with_capacity(graph.node_count()),
+            rev_node_map: HashMap::with_capacity(graph.node_count()),
+        }
+    }
+
+    fn map(&mut self) -> Vec<Swap> {
+        let num_nodes = self.graph.node_count();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // Same as digraph with no self edges
+        let mut sub_digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // A list of nodes that are remaining to be tried
+        let mut todo_nodes = Vec::with_capacity(self.input_mapping.len());
+
+        let mut rng_seed: Pcg64 = match self.rng_seed {
+            Some(rng_seed) => Pcg64::seed_from_u64(rng_seed),
+            None => Pcg64::from_entropy(),
+        };
+        // Add nodes to the digraph/sub_digraph and build the node maps
+        for node in self.graph.node_identifiers() {
+            let index = digraph.add_node(());
+            sub_digraph.add_node(());
+            self.node_map.insert(node, index);
+            self.rev_node_map.insert(index, node);
+        }
+        // The input mapping in HashMap form using NodeIndex
+        let mut tokens: HashMap<NodeIndex, NodeIndex> = self
+            .input_mapping
+            .iter()
+            .map(|(k, v)| (self.node_map[k], self.node_map[v]))
+            .collect();
+        // todo_nodes are all the mapping entries where left != right
+        for (node, dest) in &tokens {
+            if node != dest {
+                todo_nodes.push(*node);
+            }
+        }
+        // Add initial edges to the digraph/sub_digraph
+        for node in self.graph.node_identifiers() {
+            self.add_token_edges(
+                self.node_map[&node],
+                &mut digraph,
+                &mut sub_digraph,
+                &mut tokens,
+            );
+        }
+        // Do the trials and build a results Vec
+        let mut trial_results: Vec<Vec<Swap>> = Vec::new();
+        for _ in 0..self.trials {
+            let results = self.trial_map(
+                &mut digraph.clone(),
+                &mut sub_digraph.clone(),
+                &mut tokens.clone(),
+                &mut todo_nodes.clone(),
```

Comment:
Not sure how important this is. It gets a little hairy since `add_token_edges` is called before this with the uncloned graphs, etc. and it's also called from within `trial_map` on the cloned graphs.

Reply:
It's not that bad to do, I tested it locally this is the diff needed:

```diff
diff --git a/rustworkx-core/src/token_swapper.rs b/rustworkx-core/src/token_swapper.rs
index ed09648..7e7b8d9 100644
--- a/rustworkx-core/src/token_swapper.rs
+++ b/rustworkx-core/src/token_swapper.rs
@@ -128,10 +128,10 @@ where
         let mut trial_results: Vec<Vec<Swap>> = Vec::new();
         for _ in 0..self.trials {
             let results = self.trial_map(
-                &mut digraph.clone(),
-                &mut sub_digraph.clone(),
-                &mut tokens.clone(),
-                &mut todo_nodes.clone(),
+                digraph.clone(),
+                sub_digraph.clone(),
+                tokens.clone(),
+                todo_nodes.clone(),
                 &mut rng_seed,
             );
             trial_results.push(results);
@@ -202,10 +202,10 @@ where
 
     fn trial_map(
         &mut self,
-        digraph: &mut StableGraph<(), (), Directed>,
-        sub_digraph: &mut StableGraph<(), (), Directed>,
-        tokens: &mut HashMap<NodeIndex, NodeIndex>,
-        todo_nodes: &mut Vec<NodeIndex>,
+        mut digraph: StableGraph<(), (), Directed>,
+        mut sub_digraph: StableGraph<(), (), Directed>,
+        mut tokens: HashMap<NodeIndex, NodeIndex>,
+        mut todo_nodes: Vec<NodeIndex>,
         rng_seed: &mut Pcg64,
     ) -> Vec<Swap> {
         // Create a random trial list of swaps to move tokens to optimal positions
@@ -218,11 +218,11 @@ where
             let todo_node = todo_nodes[random];
 
             // If there's a cycle in sub_digraph, add it to swap_edges and do swap
-            let cycle = find_cycle(&*sub_digraph, Some(todo_node));
+            let cycle = find_cycle(&sub_digraph, Some(todo_node));
             if !cycle.is_empty() {
                 for edge in cycle[1..].iter().rev() {
                     swap_edges.push(*edge);
-                    self.swap(edge.0, edge.1, digraph, sub_digraph, tokens, todo_nodes);
+                    self.swap(edge.0, edge.1, &mut digraph, &mut sub_digraph, &mut tokens, &mut todo_nodes);
                 }
                 steps += cycle.len() - 1;
             // If there's no cycle, see if there's an edge target that matches a token key.
@@ -237,10 +237,10 @@ where
                         self.swap(
                             new_edge.0,
                             new_edge.1,
-                            digraph,
-                            sub_digraph,
-                            tokens,
-                            todo_nodes,
+                            &mut digraph,
+                            &mut sub_digraph,
+                            &mut tokens,
+                            &mut todo_nodes,
                         );
                         steps += 1;
                         found = true;
@@ -251,7 +251,7 @@ where
                 // an unhappy node. Look for a predecessor and add node and pred
                 // to swap_edges and do swap
                 if !found {
-                    let cycle: Vec<Edge> = find_cycle(&*digraph, Some(todo_node));
+                    let cycle: Vec<Edge> = find_cycle(&digraph, Some(todo_node));
                     let unhappy_node = cycle[0].0;
                     let mut found = false;
                     let di2 = &mut digraph.clone();
@@ -261,10 +261,10 @@ where
                             self.swap(
                                 unhappy_node,
                                 predecessor,
-                                digraph,
-                                sub_digraph,
-                                tokens,
-                                todo_nodes,
+                                &mut digraph,
+                                &mut sub_digraph,
+                                &mut tokens,
+                                &mut todo_nodes,
                             );
                             steps += 1;
                             found = true;
```

I think it will be good to do because it's a lot clearer about the ownership model being used. I don't think it'll have a big impact on performance or anything, but it's more idiomatic and easier to read.

Reply:
Done in https://github.com/Qiskit/rustworkx/pull/765/commits/2d74c45a8e8c6e763cd7ce32034d9f094f991661